File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-13/Estruturais/Decorator/Decorator01.js
// Componente
class ComponenteNotificador {
    enviar(msg) {
        console.log(`Enviando mensagem: ${msg}`);
    }
}

// Decorador Base
class DecoradorNotificador {
    constructor(componente) {
        this.componente = componente;
    }
    enviar(msg) {
        this.componente.enviar(msg);
    }
}

// Decorador Concreto 1 - Simulando SMS:
class NotificadorSMS extends DecoradorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por SMS: ${msg}`);
    }
}

// Decorador Concreto 2 - Simulando WhatsAop
class NotificadorWhatsApp extends DecoradorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por WhatsApp: ${msg}`);
    }
}

// Decorador Concreto 3 - Simulando Email:
class NotificadorEmail extends DecoradorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por Email: ${msg}`);
    }
}

// Interface de Utilização 
const notificador           = new ComponenteNotificador();
const notificadorSMS        = new NotificadorSMS(notificador);
const notificadorWhatsApp   = new NotificadorWhatsApp(notificador);
const notificadorEmail      = new NotificadorEmail(notificador);

const mensagem = "Ola Mundo... "
notificadorWhatsApp.enviar(mensagem);
notificadorEmail.enviar(mensagem);
notificadorSMS.enviar(mensagem);

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-09-13/Estruturais/Decorator/Decorator01.py
class ComponenteNotificador:
    def enviar(self, msg):
        print(f'Enviando Mensagem: {msg}')

class DecoradorNotificador:
    def __init__(self, componente):
        self.componente = componente
    def enviar(self, msg):
        self.componente.enviar(msg)

class NotificadorSMS(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando mensagem por SMS: {msg}')

class NotificadorWhatsApp(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando mensagem por WhatsApp: {msg}')

class NotificadorEmail(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f'Enviando mensagem por Email: {msg}')

notificador = ComponenteNotificador()
notificadorSMS =  NotificadorSMS(notificador)
notificadorWhatsApp = NotificadorWhatsApp(notificador)
notificadorEmail = NotificadorEmail(notificador)

mensagem = "Olá mundo, através do Python"
notificadorWhatsApp.enviar(mensagem)
notificadorEmail.enviar(mensagem)
notificadorSMS.enviar(mensagem)
